src/Workspaces/CSharp/Portable/LanguageServices/CSharpTypeInferenceService.TypeInferrer.cs

```diff
@@ -1465,10 +1465,19 @@ private IEnumerable<ITypeSymbol> InferTypeInPrefixUnaryExpression(PrefixUnaryExp
                     case SyntaxKind.PreIncrementExpression:
                     case SyntaxKind.UnaryPlusExpression:
                     case SyntaxKind.UnaryMinusExpression:
-                    case SyntaxKind.BitwiseNotExpression:
-                        // ++, --, +Foo(), -Foo(), ~Foo();
+                        // ++, --, +Foo(), -Foo();
                         return SpecializedCollections.SingletonEnumerable(this.Compilation.GetSpecialType(SpecialType.System_Int32));
-
+                    case SyntaxKind.BitwiseNotExpression:
+                        var binaryExpression = prefixUnaryExpression.Parent as BinaryExpressionSyntax;
+                        if (binaryExpression != null)
+                        {
+                            return InferTypeInBinaryOrAssignmentExpression(binaryExpression, binaryExpression.OperatorToken, binaryExpression.Left, binaryExpression.Right, prefixUnaryExpression);
```

Comment:
In fact, i think you should probably always calling InferTypes(prefixUnaryExpression, objectAsDefault:false), then if you get null/empty back, you pick int.  Otherwise you return what you got.